### ISSUE TITLE
"ReferenceError: weakly-referenced object no longer exists" in Pandas read_sql()

### ISSUE THREAD

**Issue by fdcastel** [pauldex/sqlalchemy-firebird#33]
**Describe the bug**
Using `read_sql()` from Pandas with the latests `sqlalchemy-firebird` raises errors which doesn't occur when using `read_sql_query()`.

**To Reproduce**
```python
import fdb
import pandas as pd
import sqlalchemy as sa

# Windows 10 x64
# Python 3.9.6
# Firebird version 3.0.7 running official Win64 build.
# sqlalchemy-firebird from 'pip install sqlalchemy-firebird'

print(pd.__version__)  # 1.3.0
print(sa.__version__)  # 1.4.25
print(fdb.__version__)  # 2.0.1

fbEngine = sa.create_engine(
    "firebird://sysdba:masterkey@localhost/C:/Program Files/Firebird/Firebird_3_0/examples/empbuild/EMPLOYEE.FDB?charset=UTF8",
    echo=True,
)

basesSql = """
SELECT
  1 "IntField",
  'Some words' "TextField",
  CURRENT_TIMESTAMP "UpdatedAt"
FROM
  mon$database
"""

with fbEngine.connect() as fbConn:
    bases = pd.read_sql_query(basesSql, fbConn)
    print(bases)
```

**Expected behavior**

```
1.3.0
1.4.25
2.0.1
2021-09-28 15:00:56,189 INFO sqlalchemy.engine.Engine
SELECT
  1 "IntField",
  'Some words' "TextField",
  CURRENT_TIMESTAMP "UpdatedAt"
FROM
  mon$database

2021-09-28 15:00:56,189 INFO sqlalchemy.engine.Engine [raw sql] ()
   IntField   TextField               UpdatedAt
0         1  Some words 2021-09-28 15:00:56.205
```

**Current behavior**

```
1.3.0
1.4.25
2.0.1
2021-09-28 15:09:36,280 INFO sqlalchemy.engine.Engine
            SELECT 1 AS has_table FROM rdb$database
            WHERE EXISTS (SELECT rdb$relation_name
                          FROM rdb$relations
                          WHERE rdb$relation_name=?)

2021-09-28 15:09:36,280 INFO sqlalchemy.engine.Engine [dialect firebird+fdb does not support caching 0.00027s] ('\nSELECT\n  1 "IntField",\n  \'Some words\' "TextField",\n  CURRENT_TIMESTAMP "UpdatedAt"\nFROM\n  mon$database\n',)
2021-09-28 15:09:36,280 INFO sqlalchemy.engine.Engine ROLLBACK
2021-09-28 15:09:36,280 INFO sqlalchemy.engine.Engine
SELECT
  1 "IntField",
  'Some words' "TextField",
  CURRENT_TIMESTAMP "UpdatedAt"
FROM
  mon$database

2021-09-28 15:09:36,280 INFO sqlalchemy.engine.Engine [raw sql] ()
Exception ignored in: <function PreparedStatement.__del__ at 0x000001E5F71F1EE0>
Traceback (most recent call last):
  File "C:\Temp\TstPandas\env\lib\site-packages\fdb\fbcore.py", line 3462, in __del__
    self._close()
  File "C:\Temp\TstPandas\env\lib\site-packages\fdb\fbcore.py", line 3338, in _close
    if is_dead_proxy(self.cursor):
  File "C:\Temp\TstPandas\env\lib\site-packages\fdb\fbcore.py", line 481, in is_dead_proxy
    return isinstance(obj, weakref.ProxyType) and not dir(obj)
ReferenceError: weakly-referenced object no longer exists
   IntField   TextField               UpdatedAt
0         1  Some words 2021-09-28 15:09:36.280
```

**Desktop (please complete the following information):**
 - OS: Windows 10 x64
 - Python 3.9.6
 - Firebird version 3.0.7 running official Win64 build.
 - sqlalchemy-firebird 0.7.0

**Additional context**
More info at: https://github.com/pauldex/sqlalchemy-firebird/issues/32

**Comment by gordthompson** [comment]
Repro without pandas:

```python
import fdb
import sqlalchemy as sa

# Python 3.9.7
# Firebird version 3.0.2 from Ubuntu 18.40 repo
# sqlalchemy-firebird from current master branch

print(sa.__version__)  # 1.4.25
print(fdb.__version__)  # 2.0.1

fbEngine = sa.create_engine(
    "firebird://sysdba:scott_tiger@localhost//home/gord/Databases/sandbox.fdb?charset=UTF8",
    echo=True,
)

basesSql = """\
SELECT 1 AS id FROM mon$database
"""

with fbEngine.connect() as fbConn:
    insp = sa.inspect(fbEngine)
    insp.has_table(basesSql)
    result = fbConn.execute(sa.text(basesSql)).fetchall()
```

pandas [documentation for read_sql()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_sql.html) says that

>This function is a convenience wrapper around `read_sql_table` and `read_sql_query` (for backward compatibility).

It calls `has_table()` to determine if it is really doing `read_sql_table()` or `read_sql_query()`. The query string is longer than 31 characters (which I assume is the maximum length for an FB table name). That causes `fdb` to fail with

> fdb.fbcore.DatabaseError: ('Cursor.fetchone:\n- SQLCODE: -303\n- Dynamic SQL Error\n- SQL error code = -303\n- arithmetic exception, numeric overflow, or string truncation\n- string right truncation\n- expected length 31, actual 33', -303, 335544569)

and then later on we get the "ReferenceError: weakly-referenced object no longer exists" thing.

TL;DR - Don't use `read_sql()`. Use `read_sql_query()` if you really are passing it a query.

**Comment by pauldex** [comment]
This is a great repro case.  The "weakly-referenced object no longer exists" message has been hard for me to track down where it happens.

**Comment by fdcastel** [comment]
@pauldex I'm unfamiliar with the code but from the (excellent) explanation given by @gordthompson couldn't this problem be fixed by just truncating the string in 31 characters (*) [here](https://github.com/pauldex/sqlalchemy-firebird/blob/master/sqlalchemy_firebird/base.py#L716)?

Or maybe put this in `denormalize_name()` (I didn't find the actual implementation code, except for a dummy one in `interfaces.py`).

(*) BTW: @gordthompson, yes, it's the table name limit for Firebird up to 3.0. It was raised to 63 in 4.0 (which should be contemplated by this code, also).

**Comment by pauldex** [comment]
That fix for Firbird 3 will break FB 4.  I'm looking for a way to tell, in that functions scope, what the maximum identifier length is at the moment of the call.

As soon as the connection is made, we can tell what version of Firebird is running (plus lots of other useful info).  But the existing parameters don't provide a way to get up to the connection object (can only execute()).

I would update to Firebird 4 if possible.  That's the quick fix.  Otherwise, I imagine I can figure out a clean way to provide FB version specific information to the places that need to make these kinds of decisions fairly soon and get them checked in.

**Comment by fdcastel** [comment]
Thank, @pauldex, for the clarification. 

Fully understood and agreed.

**Comment by pauldex** [comment]
I've checked in the fix for has_table(), which now should work correctly on all supported versions of Firebird.  Let me know if you have any other problems.
